pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,260 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package istio
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha2"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	istiov1alpha1 "knative.dev/pkg/apis/istio/common/v1alpha1"
+	istiov1alpha3 "knative.dev/pkg/apis/istio/v1alpha3"
+)
+
+const (
+	IngressConfigKeyName = "ingress"
+)
+
+// Status Constants
+var (
+	PredictorSpecMissing       = "predictorSpecMissing"
+	PredictorStatusUnknown     = "predictorStatusUnknown"
+	PredictorHostnameUnknown   = "predictorHostnameUnknown"
+	TransformerStatusUnknown   = "transformerStatusUnknown"
+	TransformerHostnameUnknown = "transformerHostnameUnknown"
+	ExplainerSpecMissing       = "explainerSpecMissing"
+	ExplainerStatusUnknown     = "explainerStatusUnknown"
+	ExplainerHostnameUnknown   = "explainerHostnameUnknown"
+)
+
+type IngressConfig struct {
+	IngressGateway     string `json:"ingressGateway,omitempty"`
+	IngressServiceName string `json:"ingressService,omitempty"`
+}
+
+type VirtualServiceBuilder struct {
+	ingressConfig *IngressConfig
+}
+
+func NewVirtualServiceBuilder(config *corev1.ConfigMap) *VirtualServiceBuilder {
+	ingressConfig := &IngressConfig{}
+	if ingress, ok := config.Data[IngressConfigKeyName]; ok {
+		err := json.Unmarshal([]byte(ingress), &ingressConfig)
+		if err != nil {
+			panic(fmt.Errorf("Unable to unmarshall ingress json string due to %v ", err))
+		}
+
+		if ingressConfig.IngressGateway == "" || ingressConfig.IngressServiceName == "" {
+			panic(fmt.Errorf("Invalid ingress config, ingressGateway and ingressService are required."))
+		}
+	}
+
+	return &VirtualServiceBuilder{ingressConfig: ingressConfig}
+}
+
+func createFailedStatus(reason string, message string) *v1alpha2.VirtualServiceStatus {
+	return &v1alpha2.VirtualServiceStatus{
+		Status: duckv1beta1.Status{
+			Conditions: duckv1beta1.Conditions{{
+				Type:    v1alpha2.RoutesReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  reason,
+				Message: message,
+			}},
+		},
+	}
+}
+
+func (r *VirtualServiceBuilder) CreateVirtualService(kfsvc *v1alpha2.KFService) (*istiov1alpha3.VirtualService, *v1alpha2.VirtualServiceStatus) {
+
+	httpRoutes := []istiov1alpha3.HTTPRoute{}
+
+	// destination for the default predict is required
+	predictDefaultSpec, defaultPredictReason := getPredictStatusConfigurationSpec(&kfsvc.Spec.Default, kfsvc.Status.Default)
+	if predictDefaultSpec == nil {
+		return nil, createFailedStatus(defaultPredictReason, "Failed to reconcile default predictor")
```

Comment:
Do we need to attach "canary" or "default" context to the error? Do we even need to determine which component it is? I think it might be cleaner to simply loop through the endpoints and then through their components, and if any of them have an error, simply surface that.

Reply:
I think its nice to have a specific error message - that said, coding this certainly made me sad :) 

In general I think our entire implementation should get redone as 2 stages:
1. resolve typed endpoints to generic endpoints with dependencies (ie a dag)
2. work the dag for status, virtual service - the lot

However that is a discussion/work for a future date :)